oauth-common/src/main/java/io/strimzi/kafka/oauth/metrics/SensorKey.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2017-2021, Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.kafka.oauth.metrics;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * The SensorKey represents and uniquely identifies the Kafka Metrics API Sensor,
+ * which represents a set of metrics that are updated at the same time.
+ * <br><br>
+ * These metrics are:
+ *
+ * <ul>
+ *     <li>count</li>
+ *     <li>totalTimeMs</li>
+ *     <li>minTimeMs</li>
+ *     <li>maxTimeMs</li>
+ *     <li>avgTimeMs</li>
+ * </ul>
+ *
+ * The SensorKey is composed of the {@code name} and the {@code attributes}.
+ * <br><br>
+ * As a convention, only use alphanumeric characters and an underscore (_) in the name and the attribute keys.
+ * Note, that names and attributes become part of the JMX ObjectName when metrics are plugged into the JMX MBeanServer.
+ * Attribute values containing some other characters (like ':', or '/') will become quoted when converted to JMX ObjectNames.
+ * <br><br>
+ * Kafka Sensors are created on demand and cached by the value of the {@code id}, which is a string representation of this SensorKey.
+ * Each Sensor is exported as a single managed bean with a unique JMX ObjectName.
+ * <br><br>
+ * Some examples of SensorKeys and how they map to JMX ObjectNames:
+ *
+ *     <ul><li> Name: <em>http_requests </em></li>
+ *         <li> Attributes: <em>context=PUBLIC, kind=jwks, host=localhost:443, path=/certs, outcome=success, status=200 </em></li>
+ *         <li> JMX ObjectName: <em>strimzi.oauth:type=http_requests,context=PUBLIC,kind=jwks,host="localhost:443",path="/certs",outcome=success,status=200</em></li>
+ *     </ul>
+ *
+ *     <ul><li> Name: <em>validation_requests </em></li>
+ *         <li> Attributes: <em>context=e8af84dd, kind=introspect, host=localhost:443, path=/introspect, outcome=error, error_type=other</em></li>
+ *         <li> JMX ObjectName: <em>strimzi.oauth:type=validation_requests,context=e8af84dd,kind=introspect,host="localhost:443",path="/introspect",outcome=error,error_type=other</em></li>
+ *     </ul>
+ */
+public class SensorKey {
+
+    final String name;
+    final Map<String, String> attributes;
+    final String id;
+
+    /**
+     * Create a new instance of SensorKey from name and a map of attributes
+     *
+     * @param name The metric name
+     * @param attributes The map of attributes
+     * @return The new instance of SensorKey
+     */
+    public static SensorKey of(String name, Map<String, String> attributes) {
+        return new SensorKey(name, attributes);
```

Comment:
If `attributes` is a `HashMap` then the order of the attributes will depend on the size of the array backing the hashmap, and therefore two `SensorKey` instances with the same (`equals`) `attributes` could have different `id`s. Similarly you could create one instance via this factory and another via the `varargs` factory and the same attributes would result in different ids. Perhaps the simplest thing would be to use a `TreeMap` in the varargs `of` and copy the `attributes` here into a `TreeMap` guaranteeing the iteration order, which should be enough to ensure the map's `toString` is also stable in the face of undefined map orderings.

Reply:
That's a fair point. Currently things have been working fine because all the usages are passing attributes as LinkedHashMap that contains a proper ordering. For the expected MBean name construction the order of the attributes is also important. I made a simple solution that should address both predictable `id` generation and expected attribute ordering for the MBean names.